File: lib/models/soap_note_v2.dart
// SOAP schema v2 model (PRD §9.3, docs/v1/05_DATA_MODEL_AND_RULES.md §5).
//
// DF-007: used for cross-client fixture tests, legacy compatibility and migration checks.
// Member screens do not read soap notes (AC-VIZ-06.1); they read memberSummaries from P2.
// Encoding and decoding live in soap_note_v2_codec.dart.
//
// Lossless rules shared by every class here:
// - A key the model does not know, or a known key whose value has the wrong type or an
//   unknown enum value, is kept in `extra` under the same key and written back as is.
//   The typed field is then null, which callers show as 'uninterpretable' (해석 불가).
// - Absent and present-null are different (V1-10 §6.1). Fields that may hold null use
//   [Present]: a Dart null field means the key is absent, `Present(null)` means the key
//   is present with null.
// - An `objective.metrics` row is either [ParsedSoapMetric] or [UnparsedSoapMetric]
//   (F-SOAP-06.2). A row is never dropped: an element that is not a map is an
//   [UnparsedSoapMetric] too, so `metrics.length` equals the stored length.
// - An `objective.snapshots` element that is not a map is kept in place as
//   [ObjectiveSnapshot.element].

import 'package:cloud_firestore/cloud_firestore.dart' show Timestamp;
import 'package:flutter/foundation.dart' show immutable;

import '../contracts/generated/metric_catalog.g.dart';
import '../contracts/generated/vocab.g.dart';

export '../contracts/generated/metric_catalog.g.dart' show MetricCode;
export '../contracts/generated/vocab.g.dart'
    show
        ActiveOrPassive,
        ChangeStatus,
        Joint,
        MdcSource,
        MetricUnit,
        Motion,
        MuscleGroup,
        ReasonCode,
        RegionCode,
        Side,
        SoapStatus,
        SourceGrade;

/// A key that is present in the document. [value] may be null (present-null).
///
/// A field of type `Present<T>?` that is itself null means the key is absent.
@immutable
final class Present<T> {
  const Present(this.value);

  final T? value;

  @override
  bool operator ==(Object other) =>
      other is Present && _deepEquals(value, other.value);

  @override
  int get hashCode => Object.hash(Present, _deepHash(value));

  @override
  String toString() => 'Present($value)';
}

/// A Firestore timestamp field: a stored instant, or the server-time placeholder of a
/// write payload (fixture tag `$serverTimestamp`).
@immutable
sealed class SoapTimestamp {
  const SoapTimestamp();

  /// A stored instant.
  const factory SoapTimestamp.instant(Timestamp timestamp) = SoapInstant;

  /// `FieldValue.serverTimestamp()` in a write payload.
  const factory SoapTimestamp.server() = SoapServerTimestamp;

  /// Convenience for hand-built models and tests.
  factory SoapTimestamp.at(DateTime time) =>
      SoapInstant(Timestamp.fromDate(time));
}

final class SoapInstant extends SoapTimestamp {
  const SoapInstant(this.timestamp);

  final Timestamp timestamp;

  DateTime toDateTime() => timestamp.toDate().toUtc();

  @override
  bool operator ==(Object other) =>
      other is SoapInstant && other.timestamp == timestamp;

  @override
  int get hashCode => timestamp.hashCode;

  @override
  String toString() => 'SoapInstant(${toDateTime().toIso8601String()})';
}

final class SoapServerTimestamp extends SoapTimestamp {
  const SoapServerTimestamp();

  @override
  bool operator ==(Object other) => other is SoapServerTimestamp;

  @override
  int get hashCode => (SoapServerTimestamp).hashCode;

  @override
  String toString() => 'SoapServerTimestamp()';
}

/// `soap_notes/{noteId}` schema v2 (V1-05 §5.1).
@immutable
final class SoapNoteV2 with _ValueEquality {
  const SoapNoteV2({
    this.schemaVersion = 2,
    this.trainerId,
    this.authorUid,
    this.memberUid,
    this.memberId,
    this.pendingMemberId,
    this.sessionDate,
    this.status,
    this.quickNote,
    this.inkPath,
    this.inkRevision,
    this.subjective,
    this.objective,
    this.exerciseAssessment,
    this.plan,
    this.memberNote,
    this.legalNature,
    this.isSharedWithMember,
    this.finalizedAt,
    this.createdAt,
    this.updatedAt,
    this.memberSummaryId,
    this.migratedFrom,
    this.legacy,
    this.extra = const {},
  });

  final int schemaVersion;
  final Present<String>? trainerId;
  final String? authorUid;
  final Present<String>? memberUid;

  /// Legacy mirror of [memberUid] during the compatibility period (P0~P3).
  final String? memberId;
  final Present<String>? pendingMemberId;
  final SoapTimestamp? sessionDate;

  /// Null when the key is absent or the value is unknown (the raw value is in [extra]).
  final SoapStatus? status;
  final String? quickNote;
  final Present<String>? inkPath;
  final int? inkRevision;
  final SoapSubjective? subjective;
  final SoapObjective? objective;
  final ExerciseAssessment? exerciseAssessment;
  final SoapPlan? plan;
  final String? memberNote;
  final String? legalNature;

  /// Legacy flag. v2 writers write false only (D9).
  final bool? isSharedWithMember;
  final SoapTimestamp? finalizedAt;
  final SoapTimestamp? createdAt;
  final SoapTimestamp? updatedAt;

  /// Server-only (P2 back reference). Read only; never written to Firestore by a client.
  final Present<String>? memberSummaryId;

  /// Server-only migration marker `{runId, schemaVersion}`. Read only.
  final Map<String, Object?>? migratedFrom;

  /// Server-only legacy originals `{metricsRaw, diagnosisRaw?, ...}`. Read only and never
  /// shown on member surfaces (MIG-04).
  final Map<String, Object?>? legacy;

  /// Unknown top-level keys and known keys whose value could not be read.
  final Map<String, Object?> extra;

  /// All `objective.metrics` rows, parsed or not.
  List<SoapMetricV2> get metrics => objective?.metrics ?? const [];

  /// Rows shown as 'uninterpretable' (해석 불가).
  int get uninterpretableMetricCount =>
      metrics.where((m) => !m.isInterpretable).length;

  @override
  List<Object?> get _props => [
        schemaVersion,
        trainerId,
        authorUid,
        memberUid,
        memberId,
        pendingMemberId,
        sessionDate,
        status,
        quickNote,
        inkPath,
        inkRevision,
        subjective,
        objective,
        exerciseAssessment,
        plan,
        memberNote,
        legalNature,
        isSharedWithMember,
        finalizedAt,
        createdAt,
        updatedAt,
        memberSummaryId,
        migratedFrom,
        legacy,
        extra,
      ];
}

/// `subjective` (V1-05 §5.2).
@immutable
final class SoapSubjective with _ValueEquality {
  const SoapSubjective({
    this.chiefComplaint,
    this.painNrs,
    this.painRegions,
    this.extra = const {},
  });

  final String? chiefComplaint;

  /// 0~10 integer. `Present(null)` means 'not entered', which is different from 0
  /// (AC-SOAP-01.7).
  final Present<int>? painNrs;

  /// Wire values as stored. Unknown codes are kept (TC-X-XC-10: reading keeps the raw value).
  final List<String>? painRegions;
  final Map<String, Object?> extra;

  /// [painRegions] parsed with the contract vocab. Unknown codes are null.
  List<RegionCode?> get painRegionCodes =>
      [for (final r in painRegions ?? const <String>[]) RegionCode.fromWire(r)];

  @override
  List<Object?> get _props => [chiefComplaint, painNrs, painRegions, extra];
}

/// `objective` (V1-05 §5.2).
@immutable
final class SoapObjective with _ValueEquality {
  const SoapObjective({
    this.metrics,
    this.refs,
    this.snapshots,
    this.extra = const {},
  });

  final List<SoapMetricV2>? metrics;
  final ObjectiveRefs? refs;
  final List<ObjectiveSnapshot>? snapshots;
  final Map<String, Object?> extra;

  @override
  List<Object?> get _props => [metrics, refs, snapshots, extra];
}

/// One `objective.metrics[]` row (V1-05 §5.2).
///
/// [ParsedSoapMetric] when the metricCode is in the catalog and every enum field is a known
/// wire value; otherwise [UnparsedSoapMetric] with the original map (F-SOAP-06.2).
@immutable
sealed class SoapMetricV2 {
  const SoapMetricV2();

  const factory SoapMetricV2.parsed({
    required MetricCode metricCode,
    required num value,
    required MetricUnit unit,
    required Side side,
    required SourceGrade sourceGrade,
    Joint? joint,
    Motion? motion,
    ActiveOrPassive? activeOrPassive,
    MuscleGroup? muscleGroup,
    String? note,
    Map<String, Object?> extra,
  }) = ParsedSoapMetric;

  const factory SoapMetricV2.unparsed(Object? raw) = UnparsedSoapMetric;

  /// False for rows shown as 'uninterpretable' (해석 불가).
  bool get isInterpretable;
}

final class ParsedSoapMetric extends SoapMetricV2 with _ValueEquality {
  const ParsedSoapMetric({
    required this.metricCode,
    required this.value,
    required this.unit,
    required this.side,
    required this.sourceGrade,
    this.joint,
    this.motion,
    this.activeOrPassive,
    this.muscleGroup,
    this.note,
    this.extra = const {},
  });

  final MetricCode metricCode;

  /// `mmtGrade` values are integers (fixture tag `$int`); other values are doubles.
  final num value;
  final MetricUnit unit;
  final Side side;
  final SourceGrade sourceGrade;
  final Joint? joint;
  final Motion? motion;
  final ActiveOrPassive? activeOrPassive;
  final MuscleGroup? muscleGroup;
  final String? note;

  /// Row keys this version does not know. Kept and written back.
  final Map<String, Object?> extra;

  @override
  bool get isInterpretable => true;

  MetricCatalogEntry get catalogEntry => MetricCatalog.entry(metricCode);

  /// Whether the row meets V1-05 §5.2 (catalog unit, sourceGrade, side rule and the
  /// romDeg/mmtGrade condition fields). Incomplete rows are shown as '미완성' and are
  /// not saved on finalize (F-SOAP-02.6). The codec keeps them either way.
  bool get isComplete {
    final entry = catalogEntry;
    if (entry.storage.field != 'objective.metrics') return false;
    if (unit != entry.unit) return false;
    if (!entry.allowedSourceGrades.contains(sourceGrade)) return false;
    switch (entry.sideRule) {
      case SideRule.leftRightBilateral:
        if (side == Side.none) return false;
      case SideRule.leftRight:
        if (side != Side.left && side != Side.right) return false;
      case SideRule.none:
        if (side != Side.none) return false;
      case SideRule.magnitudeWithLowerSide ||
            SideRule.magnitudeWithHigherSide ||
            SideRule.undefinedV2:
        break;
    }
    switch (metricCode) {
      case MetricCode.romDeg:
        return joint != null && motion != null && activeOrPassive != null;
      case MetricCode.mmtGrade:
        return muscleGroup != null && value is int && value >= 0 && value <= 5;
      default:
        return true;
    }
  }

  @override
  List<Object?> get _props => [
        metricCode,
        value,
        unit,
        side,
        sourceGrade,
        joint,
        motion,
        activeOrPassive,
        muscleGroup,
        note,
        extra,
      ];
}

final class UnparsedSoapMetric extends SoapMetricV2 with _ValueEquality {
  const UnparsedSoapMetric(this.raw);

  /// The original row, written back unchanged. Normally a map; any other stored element
  /// (a string, a number, null, a list) is kept as it is so the row count never shrinks.
  final Object? raw;

  /// Whether the stored row is a map.
  bool get isMapRow => raw is Map;

  /// The raw `metricCode`, if the row is a map and the code is a string.
  String? get metricCodeWire {
    final row = raw;
    final code = row is Map ? row['metricCode'] : null;
    return code is String ? code : null;
  }

  @override
  bool get isInterpretable => false;

  @override
  List<Object?> get _props => [raw];
}

/// `objective.refs` (V1-05 §5.2).
@immutable
final class ObjectiveRefs with _ValueEquality {
  const ObjectiveRefs({
    this.postureAssessmentIds,
    this.bodyCompositionRecordIds,
    this.circumferenceMeasurementIds,
    this.bodyScanIds,
    this.extra = const {},
  });

  final List<String>? postureAssessmentIds;
  final List<String>? bodyCompositionRecordIds;
  final List<String>? circumferenceMeasurementIds;
  final List<String>? bodyScanIds;
  final Map<String, Object?> extra;

  @override
  List<Object?> get _props => [
        postureAssessmentIds,
        bodyCompositionRecordIds,
        circumferenceMeasurementIds,
        bodyScanIds,
        extra,
      ];
}

/// `objective.snapshots[]`: immutable copy of a referenced record value (PRD §7.8).
@immutable
final class ObjectiveSnapshot with _ValueEquality {
  const ObjectiveSnapshot({
    this.element,
    this.refId,
    this.metricCode,
    this.value,
    this.unit,
    this.side,
    this.sourceGrade,
    this.changeStatus,
    this.reasonCode,
    this.policyVersion,
    this.mdcSource,
    this.measuredAt,
    this.extra = const {},
  });

  final String? refId;
  final MetricCode? metricCode;
  final num? value;
  final MetricUnit? unit;
  final Side? side;
  final SourceGrade? sourceGrade;

  /// Always `pendingPolicy` before P3 (ADR-009).
  final ChangeStatus? changeStatus;
  final Present<ReasonCode>? reasonCode;
  final Present<String>? policyVersion;
  final Present<MdcSource>? mdcSource;
  final SoapTimestamp? measuredAt;
  final Map<String, Object?> extra;

  /// Set only when the stored element is not a map: the element as stored (possibly
  /// `Present(null)`), written back unchanged. Every other field is then null.
  final Present<Object>? element;

  /// Whether the stored element was a map that could be read field by field.
  bool get isReadable => element == null;

  @override
  List<Object?> get _props => [
        element,
        refId,
        metricCode,
        value,
        unit,
        side,
        sourceGrade,
        changeStatus,
        reasonCode,
        policyVersion,
        mdcSource,
        measuredAt,
        extra,
      ];
}

/// `exerciseAssessment` (V1-05 §5.1).
@immutable
final class ExerciseAssessment with _ValueEquality {
  const ExerciseAssessment({
    this.summary,
    this.observations,
    this.extra = const {},
  });

  final String? summary;
  final List<String>? observations;
  final Map<String, Object?> extra;

  @override
  List<Object?> get _props => [summary, observations, extra];
}

/// `plan` (V1-05 §5.1).
@immutable
final class SoapPlan with _ValueEquality {
  const SoapPlan({this.nextSession, this.homeExercise, this.extra = const {}});

  final String? nextSession;
  final String? homeExercise;
  final Map<String, Object?> extra;

  @override
  List<Object?> get _props => [nextSession, homeExercise, extra];
}

// -------------------------------------------------------------------------------------------
// Value equality. Maps compare without key order; `null` and a missing key are different.
// -------------------------------------------------------------------------------------------

mixin _ValueEquality {
  List<Object?> get _props;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      (other.runtimeType == runtimeType &&
          other is _ValueEquality &&
          _deepEquals(_props, other._props));

  @override
  int get hashCode => _deepHash(_props);

  @override
  String toString() => '$runtimeType$_props';
}

bool _deepEquals(Object? a, Object? b) {
  if (identical(a, b)) return true;
  if (a is Map && b is Map) {
    if (a.length != b.length) return false;
    for (final key in a.keys) {
      if (!b.containsKey(key) || !_deepEquals(a[key], b[key])) return false;
    }
    return true;
  }
  if (a is List && b is List) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (!_deepEquals(a[i], b[i])) return false;
    }
    return true;
  }
  return a == b;
}

int _deepHash(Object? v) {
  if (v is Map) {
    var h = 0;
    for (final e in v.entries) {
      h ^= Object.hash(e.key, _deepHash(e.value));
    }
    return Object.hash(Map, h, v.length);
  }
  if (v is List) return Object.hashAll(v.map(_deepHash));
  return v.hashCode;
}

File: trainer_app/Packages/TrainerCore/Sources/TrainerDomain/SoapNoteV2.swift
import Foundation
import TrainerContracts

// SOAP schema v2 domain model (DF-009, PRD §9.3, docs/v1/05_DATA_MODEL_AND_RULES.md §5).
// Encoding and decoding live in SoapNoteV2Codec.swift. The Dart twin is lib/models/soap_note_v2.dart (DF-007);
// both read and write the same contracts/fixtures/soap_v2 documents with the same meaning (NFR-07).
//
// Lossless rules shared by every type here:
// - A key the model does not know, or a known key whose value has the wrong type or an unknown enum value,
//   is kept in `extra` under the same key and written back as is. The typed property is then nil, which
//   screens show as 'uninterpretable' (해석 불가).
// - Absent and present-null are different (V1-10 §6.1). Properties that may hold null use `Nullable`:
//   a Swift `nil` means the key is absent, `.null` means the key is present with null.
// - An `objective.metrics` row is `.parsed` or `.unparsed(raw:)` (F-SOAP-06.2). A row is never dropped,
//   and an empty list stays empty: nothing fills it with demo text (F-SOAP-06.3).
// - Enums are the generated contract enums (TrainerContracts, English wire values only). Legacy Korean
//   notations never parse (SoapModels.swift:4-8 counterexample); legacy documents use `LegacySoapView`.

/// A key that is present in the document and may hold null. A property of type `Nullable<T>?` that is
/// itself nil means the key is absent.
public enum Nullable<Wrapped> {
  case null
  case value(Wrapped)

  /// The wrapped value, or nil for `.null`.
  public var value: Wrapped? {
    if case let .value(v) = self { return v }
    return nil
  }
}

extension Nullable: Equatable where Wrapped: Equatable {}
extension Nullable: Hashable where Wrapped: Hashable {}
extension Nullable: Sendable where Wrapped: Sendable {}

/// A Firestore timestamp field: a stored instant, or the server-time placeholder of a write payload.
public enum SoapTimestamp: Equatable, Hashable, Sendable {
  case instant(Date)
  case server
}

/// Whose record this is (V1-05 §3.3: `memberKey = memberUid ?? pendingMemberId`).
///
/// Encoded as `memberUid` and `pendingMemberId` (exactly one holds a value at create). A `.uid` also
/// writes the legacy `memberId` mirror with the same value during the compatibility period (§9.3, P0~P3).
public enum MemberKey: Hashable, Sendable {
  case uid(String)
  case pending(String)

  /// The key value (`memberUid` or `pendingMemberId`).
  public var id: String {
    switch self {
    case let .uid(id), let .pending(id): return id
    }
  }
}

/// `soap_notes/{noteId}` schema v2 (V1-05 §5.1).
public struct SoapNoteV2: Equatable, Sendable {
  /// Always 2. Documents without it are v1 and are read with `LegacySoapView` (V1-05 §5.6).
  public var schemaVersion: Int { 2 }

  public var trainerId: Nullable<String>?
  public var authorUid: String?

  /// Nil when the stored `memberUid`/`pendingMemberId`/`memberId` combination is not a v2 shape; the
  /// stored keys are then kept in `extra` unchanged.
  public var member: MemberKey?
  /// `.uid` only: `pendingMemberId` that a promoted record keeps next to `memberUid` (V1-05 §3.3).
  /// Nil writes `pendingMemberId: null`.
  public var promotedFromPendingMemberId: String?
  /// `.uid` only: whether the legacy `memberId` mirror is written (V1-05 §5.1, optional, P0~P3).
  public var writesLegacyMemberId: Bool

  public var sessionDate: SoapTimestamp?
  /// Nil when the key is absent or the value is unknown (the raw value is then in `extra`).
  public var status: SoapStatus?
  /// The Live one-liner, kept verbatim (F-SOAP-02.1).
  public var quickNote: String?
  public var inkPath: Nullable<String>?
  public var inkRevision: Int64?
  public var subjective: SoapSubjective?
  public var objective: SoapObjective?
  public var exerciseAssessment: ExerciseAssessment?
  public var plan: SoapPlan?
  public var memberNote: String?
  public var legalNature: String?
  /// Legacy flag. v2 writers write false only (D9).
  public var isSharedWithMember: Bool?
  public var finalizedAt: SoapTimestamp?
  public var createdAt: SoapTimestamp?
  public var updatedAt: SoapTimestamp?

  /// Server-only (P2 back reference). Read only; the Firestore target never writes it.
  public var memberSummaryId: Nullable<String>?
  /// Server-only migration marker `{runId, schemaVersion}`. Read only.
  public var migratedFrom: [String: JSONValue]?
  /// Server-only legacy originals `{metricsRaw, diagnosisRaw?, ...}`. Read only, never shown to members (MIG-04).
  public var legacy: [String: JSONValue]?

  /// Unknown top-level keys and known keys whose value could not be read.
  public var extra: [String: JSONValue]

  public init(
    trainerId: Nullable<String>? = nil,
    authorUid: String? = nil,
    member: MemberKey? = nil,
    promotedFromPendingMemberId: String? = nil,
    writesLegacyMemberId: Bool = true,
    sessionDate: SoapTimestamp? = nil,
    status: SoapStatus? = nil,
    quickNote: String? = nil,
    inkPath: Nullable<String>? = nil,
    inkRevision: Int64? = nil,
    subjective: SoapSubjective? = nil,
    objective: SoapObjective? = nil,
    exerciseAssessment: ExerciseAssessment? = nil,
    plan: SoapPlan? = nil,
    memberNote: String? = nil,
    legalNature: String? = nil,
    isSharedWithMember: Bool? = nil,
    finalizedAt: SoapTimestamp? = nil,
    createdAt: SoapTimestamp? = nil,
    updatedAt: SoapTimestamp? = nil,
    memberSummaryId: Nullable<String>? = nil,
    migratedFrom: [String: JSONValue]? = nil,
    legacy: [String: JSONValue]? = nil,
    extra: [String: JSONValue] = [:]
  ) {
    self.trainerId = trainerId
    self.authorUid = authorUid
    self.member = member
    self.promotedFromPendingMemberId = promotedFromPendingMemberId
    self.writesLegacyMemberId = writesLegacyMemberId
    self.sessionDate = sessionDate
    self.status = status
    self.quickNote = quickNote
    self.inkPath = inkPath
    self.inkRevision = inkRevision
    self.subjective = subjective
    self.objective = objective
    self.exerciseAssessment = exerciseAssessment
    self.plan = plan
    self.memberNote = memberNote
    self.legalNature = legalNature
    self.isSharedWithMember = isSharedWithMember
    self.finalizedAt = finalizedAt
    self.createdAt = createdAt
    self.updatedAt = updatedAt
    self.memberSummaryId = memberSummaryId
    self.migratedFrom = migratedFrom
    self.legacy = legacy
    self.extra = extra
  }

  /// All `objective.metrics` rows, parsed or not, in stored order.
  public var metrics: [SoapMetricV2] { objective?.metrics ?? [] }

  /// Rows shown as 'uninterpretable' (해석 불가).
  public var uninterpretableMetricCount: Int { metrics.filter { !$0.isInterpretable }.count }
}

/// `subjective` (V1-05 §5.2).
public struct SoapSubjective: Equatable, Sendable {
  public var chiefComplaint: String?
  /// 0~10 integer. `.null` means 'not entered', which is different from 0 (AC-SOAP-01.7).
  public var painNrs: Nullable<Int64>?
  /// Wire values as stored. Unknown codes are kept (reading keeps the raw value).
  public var painRegions: [String]?
  public var extra: [String: JSONValue]

  public init(
    chiefComplaint: String? = nil, painNrs: Nullable<Int64>? = nil, painRegions: [String]? = nil,
    extra: [String: JSONValue] = [:]
  ) {
    self.chiefComplaint = chiefComplaint
    self.painNrs = painNrs
    self.painRegions = painRegions
    self.extra = extra
  }

  /// `painRegions` read with the contract vocab. Unknown codes are nil.
  public var painRegionCodes: [RegionCode?] { (painRegions ?? []).map { RegionCode(rawValue: $0) } }
}

/// `objective` (V1-05 §5.2).
public struct SoapObjective: Equatable, Sendable {
  public var metrics: [SoapMetricV2]?
  public var refs: ObjectiveRefs?
  public var snapshots: [ObjectiveSnapshot]?
  public var extra: [String: JSONValue]

  public init(
    metrics: [SoapMetricV2]? = nil, refs: ObjectiveRefs? = nil, snapshots: [ObjectiveSnapshot]? = nil,
    extra: [String: JSONValue] = [:]
  ) {
    self.metrics = metrics
    self.refs = refs
    self.snapshots = snapshots
    self.extra = extra
  }
}

/// One `objective.metrics[]` row (V1-05 §5.2).
///
/// `.parsed` when the metricCode is in the catalog and every enum field is a known wire value;
/// otherwise `.unparsed(raw:)` with the stored element, written back unchanged (F-SOAP-06.2).
public enum SoapMetricV2: Equatable, Sendable {
  case parsed(ParsedMetric)
  case unparsed(raw: JSONValue)

  /// False for rows shown as 'uninterpretable' (해석 불가).
  public var isInterpretable: Bool {
    if case .parsed = self { return true }
    return false
  }

  /// The raw `metricCode` of an unparsed object row whose code is a string.
  public var unparsedMetricCode: String? {
    if case let .unparsed(raw) = self { return raw["metricCode"]?.stringValue }
    return nil
  }
}

/// A metric row whose code and enum fields this version knows.
public struct ParsedMetric: Equatable, Sendable {
  public var metricCode: MetricCode
  /// `mmtGrade` values are integral and are written as integers (fixture tag `$int`); others as doubles.
  public var value: Double
  public var unit: MetricUnit
  public var side: Side
  public var sourceGrade: SourceGrade
  public var joint: Joint?
  public var motion: Motion?
  public var activeOrPassive: ActiveOrPassive?
  public var muscleGroup: MuscleGroup?
  public var note: String?
  /// Row keys this version does not know. Kept and written back.
  public var extra: [String: JSONValue]

  public init(
    metricCode: MetricCode, value: Double, unit: MetricUnit, side: Side, sourceGrade: SourceGrade,
    joint: Joint? = nil, motion: Motion? = nil, activeOrPassive: ActiveOrPassive? = nil,
    muscleGroup: MuscleGroup? = nil, note: String? = nil, extra: [String: JSONValue] = [:]
  ) {
    self.metricCode = metricCode
    self.value = value
    self.unit = unit
    self.side = side
    self.sourceGrade = sourceGrade
    self.joint = joint
    self.motion = motion
    self.activeOrPassive = activeOrPassive
    self.muscleGroup = muscleGroup
    self.note = note
    self.extra = extra
  }

  public var catalogEntry: MetricCatalogEntry { MetricCatalog.entry(for: metricCode) }

  /// Whether the row meets V1-05 §5.2 (catalog unit, sourceGrade, side rule and the romDeg/mmtGrade
  /// condition fields). Incomplete rows are shown as '미완성' and are not saved on finalize (F-SOAP-02.6).
  /// The codec keeps them either way. Same rule as the Dart `ParsedSoapMetric.isComplete`.
  public var isComplete: Bool {
    let entry = catalogEntry
    guard entry.storage.field == "objective.metrics", unit == entry.unit,
      entry.allowedSourceGrades.contains(sourceGrade)
    else { return false }
    switch entry.sideRule {
    case .leftRightBilateral: if side == .none { return false }
    case .leftRight: if side != .left && side != .right { return false }
    case .none: if side != .none { return false }
    case .magnitudeWithLowerSide, .magnitudeWithHigherSide, .undefinedV2: break
    }
    switch metricCode {
    case .romDeg: return joint != nil && motion != nil && activeOrPassive != nil
    case .mmtGrade: return muscleGroup != nil && value == value.rounded() && (0...5).contains(value)
    default: return true
    }
  }
}

/// `objective.refs` (V1-05 §5.2).
public struct ObjectiveRefs: Equatable, Sendable {
  public var postureAssessmentIds: [String]?
  public var bodyCompositionRecordIds: [String]?
  public var circumferenceMeasurementIds: [String]?
  public var bodyScanIds: [String]?
  public var extra: [String: JSONValue]

  public init(
    postureAssessmentIds: [String]? = nil, bodyCompositionRecordIds: [String]? = nil,
    circumferenceMeasurementIds: [String]? = nil, bodyScanIds: [String]? = nil, extra: [String: JSONValue] = [:]
  ) {
    self.postureAssessmentIds = postureAssessmentIds
    self.bodyCompositionRecordIds = bodyCompositionRecordIds
    self.circumferenceMeasurementIds = circumferenceMeasurementIds
    self.bodyScanIds = bodyScanIds
    self.extra = extra
  }

  /// All four lists present and empty: what a new note writes.
  public static let empty = ObjectiveRefs(
    postureAssessmentIds: [], bodyCompositionRecordIds: [], circumferenceMeasurementIds: [], bodyScanIds: [])
}

/// `objective.snapshots[]`: immutable copy of a referenced record value (PRD §7.8).
public struct ObjectiveSnapshot: Equatable, Sendable {
  public var refId: String?
  /// When nil, a stored `value` is kept in `extra` as read.
  public var metricCode: MetricCode?
  public var value: Double?
  public var unit: MetricUnit?
  public var side: Side?
  public var sourceGrade: SourceGrade?
  /// Always `pendingPolicy` before P3 (ADR-009).
  public var changeStatus: ChangeStatus?
  public var reasonCode: Nullable<ReasonCode>?
  public var policyVersion: Nullable<String>?
  public var mdcSource: Nullable<MdcSource>?
  public var measuredAt: SoapTimestamp?
  public var extra: [String: JSONValue]
  /// Set only when the stored element is not an object: the element as stored (possibly `.null`),
  /// written back unchanged. Every other property is then nil.
  public var element: JSONValue?

  public init(
    refId: String? = nil, metricCode: MetricCode? = nil, value: Double? = nil, unit: MetricUnit? = nil,
    side: Side? = nil, sourceGrade: SourceGrade? = nil, changeStatus: ChangeStatus? = nil,
    reasonCode: Nullable<ReasonCode>? = nil, policyVersion: Nullable<String>? = nil,
    mdcSource: Nullable<MdcSource>? = nil, measuredAt: SoapTimestamp? = nil, extra: [String: JSONValue] = [:],
    element: JSONValue? = nil
  ) {
    self.refId = refId
    self.metricCode = metricCode
    self.value = value
    self.unit = unit
    self.side = side
    self.sourceGrade = sourceGrade
    self.changeStatus = changeStatus
    self.reasonCode = reasonCode
    self.policyVersion = policyVersion
    self.mdcSource = mdcSource
    self.measuredAt = measuredAt
    self.extra = extra
    self.element = element
  }

  /// Whether the stored element was an object that could be read field by field.
  public var isReadable: Bool { element == nil }
}

/// `exerciseAssessment` (V1-05 §5.1).
public struct ExerciseAssessment: Equatable, Sendable {
  public var summary: String?
  public var observations: [String]?
  public var extra: [String: JSONValue]

  public init(summary: String? = nil, observations: [String]? = nil, extra: [String: JSONValue] = [:]) {
    self.summary = summary
    self.observations = observations
    self.extra = extra
  }
}

/// `plan` (V1-05 §5.1).
public struct SoapPlan: Equatable, Sendable {
  public var nextSession: String?
  public var homeExercise: String?
  public var extra: [String: JSONValue]

  public init(nextSession: String? = nil, homeExercise: String? = nil, extra: [String: JSONValue] = [:]) {
    self.nextSession = nextSession
    self.homeExercise = homeExercise
    self.extra = extra
  }
}
